finagle/h2/src/main/scala/com/twitter/finagle/buoyant/h2/netty4/Netty4StreamTransport.scala

```diff
@@ -306,208 +211,170 @@ private[h2] trait Netty4StreamTransport[SendMsg <: Message, RecvMsg <: Message]
-          case Closed(_) => false
-
-          case RemoteOpen(remote) =>
-            if (resetFromRemote(remote, err)) {
-              statsReceiver.remoteResetCount.incr()
-              true
-            } else recv(rst)
-
-          case state@RemoteClosed() =>
-            if (resetFromRemote(state, err)) {
-              statsReceiver.remoteResetCount.incr()
-              true
-            } else recv(rst)
-        }
+        if (reset(err, local = false)) {
+          statsReceiver.remoteResetCount.incr()
+          true
+        } else false
 
+      /*
+       * HEADERS FRAME with EOS (TRAILERS)
+       */
       case hdrs: Http2HeadersFrame if hdrs.isEndStream =>
+        //Such a Frame may be received while in RecvPending (headers
+        // only message) or while in RecvStreaming (trailers).
         state match {
-          case Closed(_) => false
-
-          case state@RemoteClosed() =>
-            if (resetFromLocal(state, Reset.InternalError)) true
-            else recv(hdrs)
-
-          case Open(remote@RemotePending()) =>
-            val q = new AsyncQueue[Frame](1)
-            val msg = mkRecvMsg(hdrs.headers, Stream.empty(q))
-            if (ConnectionHeaders.detect(msg.headers)) {
-              if (resetFromLocal(remote, Reset.ProtocolError)) true
-              else recv(hdrs)
-            } else {
-              if (stateRef.compareAndSet(state, new RemoteClosed(q))) {
-                remote.setMessage(msg)
-                true
-              } else recv(hdrs)
-            }
-
-          case Open(remote@RemoteStreaming()) =>
-            if (stateRef.compareAndSet(state, remote.toRemoteClosed)) {
-              if (recvFrame(toFrame(hdrs), remote)) {
-                remote.close()
-                true
-              } else false
-            } else recv(hdrs)
-
-          case state@LocalClosed(remote@RemotePending()) =>
-            val msg = mkRecvMsg(hdrs.headers, NilStream)
+          // Pending
+          case StreamState(sendState, RecvPending) =>
+            val stream = Stream.empty()
+            val msg = mkRecvMsg(hdrs.headers, stream)
             if (ConnectionHeaders.detect(msg.headers)) {
-              if (resetFromLocal(state, Reset.ProtocolError)) true
-              else recv(hdrs)
+              log.debug("[%s] illegal connection headers detected in %s", prefix, msg.headers)
+              reset(Reset.ProtocolError, local = true)
+              false
             } else {
-              if (stateRef.compareAndSet(state, Closed(Reset.NoError))) {
-                remote.setMessage(msg)
-                resetP.setDone()
+              if (stateRef.compareAndSet(state, StreamState(sendState, RecvClosed))) {
+                stream.onCancel.onSuccess { rst =>
+                  // If the recvMsg stream is cancelled, reset the stream.
+                  reset(rst, local = true); ()
+                }
+                recvMsg.setValue(msg)
+                if (sendState == SendClosed) {
```

Comment:
i imagine we don't actually need to set the onCancel handler when sendState is SendClosed? (because this basically means we're in full-close///